src/constbuffer.c

```diff
@@ -394,3 +398,256 @@ IMPLEMENT_MOCKABLE_FUNCTION(, bool, CONSTBUFFER_HANDLE_contain_same, CONSTBUFFER
+        (source == NULL) ||
+        /*Codes_SRS_CONSTBUFFER_02_056: [ If destination is NULL then CONSTBUFFER_to_fixed_size_buffer shall fail and return CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_INVALID_ARG. ]*/
+        (destination == NULL) ||
+        /*Codes_SRS_CONSTBUFFER_02_057: [ If serialized_size is NULL then CONSTBUFFER_to_fixed_size_buffer shall fail and return CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_INVALID_ARG. ]*/
+        (serialized_size == NULL)
+        )
+    {
+        LogError("invalid arguments CONSTBUFFER_HANDLE source=%p, unsigned char* destination=%p, uint32_t destination_size=%" PRIu32 ", uint32_t* serialized_size=%p",
+            source, destination, destination_size, serialized_size);
+        result = CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_INVALID_ARG;
+    }
+    else
+    {
+        if (UINT32_MAX - (sizeof(uint8_t) + sizeof(uint32_t)) < source->alias.size)
+        {
+            /*when this is true, an output for "serialized_size" cannot be provided as the complete serialization would exceed UINT32_MAX*/
+
+            /*Codes_SRS_CONSTBUFFER_02_074: [ If there are any failures then CONSTBUFFER_to_fixed_size_buffer shall fail and return CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_ERROR. ]*/
+            result = CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_ERROR;
+        }
+        else
+        {
+            *serialized_size = sizeof(uint8_t) + sizeof(uint32_t) + (uint32_t)source->alias.size;
+            /*Codes_SRS_CONSTBUFFER_02_058: [ If the size of serialization exceeds destination_size then CONSTBUFFER_to_fixed_size_buffer shall fail, write in serialized_size how much it would need and return CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_INSUFFICIENT_BUFFER. ]*/
+            if (destination_size < sizeof(uint8_t) + sizeof(uint32_t))
+            {
+                LogError("destination=%p does not contain enough bytes for version and size. It only has %" PRIu32 " bytes", destination, destination_size);
+                result = CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_INSUFFICIENT_BUFFER;
+            }
+            else
+            {
+                if (destination_size - (sizeof(uint8_t) + sizeof(uint32_t)) < source->alias.size)
+                {
+                    /*Codes_SRS_CONSTBUFFER_02_058: [ If the size of serialization exceeds destination_size then CONSTBUFFER_to_fixed_size_buffer shall fail, write in serialized_size how much it would need and return CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_INSUFFICIENT_BUFFER. ]*/
+                    LogError("cannot serialize CONSTBUFFER_HANDLE source=%p because its serialized size (sizeof(uint8_t) + sizeof(uint32_t) + alias->size=%zu)=%zu exceeds available destination_size=%" PRIu32 "",
+                        source, source->alias.size, sizeof(uint8_t) + sizeof(uint32_t) + source->alias.size, destination_size);
+                    result = CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_INSUFFICIENT_BUFFER;
+                }
```

Comment:
Why do we need 2 checks? I think we only need one: we know what is the serialized size, if the destimation buffer size is less we should error, printing the sizes and that should be it.

Reply:
Correct. I was too pedantic about verifying at every step.